src/main/java/com/derongan/minecraft/deeperworld/listeners/MovementListener.kt

```diff
@@ -81,10 +101,100 @@ object MovementListener : Listener {
-        player.fallDistance = fallDistance
-        player.velocity = oldVelocity
+
+        val oldLeashedEntities = player.getLeashedEntities()
+        oldLeashedEntities?.forEach {
+            if (it.customName.isNullOrEmpty()) {
+                it.customName = " "
+            }
+        }
+
+        val rootVehicle = player.getVehicleRecursive()
+        if (rootVehicle != null) {
+            newLoc.yaw = player.location.yaw
+            newLoc.pitch = player.location.pitch
+
+            // Prevent teleportation of other players in the vehicle-passenger structure.
+            rootVehicle.getPlayerPassengersRecursive()?.forEach {
+                if (it != player) {
+                    it.vehicle?.removePassenger(it)
+                }
+            }
+
+            val oldFallDistance = rootVehicle.fallDistance
+            val oldVelocity = rootVehicle.velocity
+
+            val vehicleTree = VehicleTree(rootVehicle)
+
+            vehicleTree.root.applyAll {
+                it.value.passengers.forEach { passenger ->
+                    it.value.removePassenger(passenger)
+                }
+
+                if (it.value.customName.isNullOrEmpty()) {
+                    it.value.customName = " "
+                }
+            }
+
+            player.teleport(newLoc)
+
+            Bukkit.getServer().scheduler.schedule(deeperWorld) {
+                waitFor(DeeperConfig.data.entityTeleportDelay)
+
+                vehicleTree.root.applyAll {
+                    it.value.teleport(player)
+                    if (it.value.customName == " ") {
+                        it.value.customName = null
+                    }
+                }
+
+                vehicleTree.root.applyAll { vehicleNode ->
+                    vehicleNode.children.forEach {
+                        vehicleNode.value.addPassenger(it.value)
+                    }
+                }
+
+                rootVehicle.fallDistance = oldFallDistance
+                rootVehicle.velocity = oldVelocity
+
+                oldLeashedEntities?.forEach { leashEntity ->
+
+                    // Delete the lead that drops when the player teleports and unleashes the entity
+                    leashEntity.getNearbyItemEntities(20.0, 20.0, 20.0, Material.LEAD)
+                            ?.minByOrNull { it.location.distanceSquared(leashEntity.location) }
+                            ?.remove()
+
+                    leashEntity.teleport(player)
+                    leashEntity.setLeashHolder(player)
+                    if (leashEntity.customName == " ") {
+                        leashEntity.customName = null
+                    }
+                }
+            }
+        } else {
+            val fallDistance = player.fallDistance
+            val oldVelocity = player.velocity
+            player.teleport(newLoc)
+            player.fallDistance = fallDistance
+            player.velocity = oldVelocity
+
+            Bukkit.getServer().scheduler.schedule(deeperWorld) {
```

Comment:
Don't like this repeated code, probably worth moving into a private function. If you need access to some extra context, you can put a function within this function.

Reply:
With the other changes, the repeated code is reduced to this:

```
oldLeashedEntities.forEach { leashEntity ->
                    leashEntity.teleport(player)
                    leashEntity.setLeashHolder(player)
                }
```

Don't think a private function is necessary.